packages/react-devtools-shared/src/backend/fiber/renderer.js

```diff
@@ -2419,6 +2423,25 @@ export function attach(
     pendingRealUnmountedIDs.push(id);
   }
 
+  function getSecondaryEnvironmentName(
+    debugInfo: ?ReactDebugInfo,
+    index: number,
+  ): null | string {
+    if (debugInfo != null) {
+      const componentInfo: ReactComponentInfo = (debugInfo[index]: any);
+      for (let i = index + 1; i < debugInfo.length; i++) {
+        const debugEntry = debugInfo[i];
+        if (typeof debugEntry.env === 'string') {
+          // If the next environment is different then this component was the boundary
+          // and it changed before entering the next component. So we assign this
+          // component a secondary environment.
+          return componentInfo.env !== debugEntry.env ? debugEntry.env : null;
```

Comment:
What if `componentInfo.env` is nullish, but `debugEntry.env` is not? Like when VirtualInstance is a descendant of FiberInstance.

Reply:
There's no instances here. This is just walking the debugInfo array and comparing it to other entires in that array.

`componentInfo.env` really can't be nullish right now since there's always an environment for a server component. It's mainly just defensive in case we support other types of virtual components in the future. If it differs, then it's a change in environment so should specify the secondary if available.

However, I don't actually use the secondary environment later on if the first one isn't specified. I assume that then it was just a client component or something that was inlined. But that case doesn't really exist yet.